msal/application.py

```diff
@@ -763,13 +765,22 @@ def _build_client(self, client_credential, authority, skip_regional_client=False
                     raise ValueError(
                         "client_credential needs to follow this format "
                         "https://msal-python.readthedocs.io/en/latest/#msal.ClientApplication.params.client_credential")
-                if ("x5c" not in headers  # So we did not run the pfx code path
+                if ("x5c" not in headers  # So the .pfx file contains no certificate
                     and isinstance(client_credential.get('public_certificate'), str)
                 ):  # Then we treat the public_certificate value as PEM content
                     headers["x5c"] = extract_certs(client_credential['public_certificate'])
+                if sha256_thumbprint and not authority.is_adfs:
+                    assertion_params = {
+                        "algorithm": "PS256", "sha256_thumbprint": sha256_thumbprint,
+                    }
+                else:  # Fall back
+                    if not sha1_thumbprint:
+                        raise ValueError("You shall provide a thumbprint in SHA1.")
```

Comment:
Are the SHA thumbprints something that the developer is meant to provide, and would they see this error message if something went wrong? If so, you may want something more informative like "A SHA1 thumbprint is needed for ADFS" or similar to let them know which SHA version is needed for each flow.

Reply:
When an app developer encounters this error message, he or she is already in the ADFS code path, so the "you shall provide a thumbprint in SHA1" is good enough as an actionable item.

I incline to avoid mentioning too-specific info such as "ADFS" in this case, because that kind of wording tends to become outdated when/if we will add or remove some scenarios into this SHA1 group.